### PR TITLE
fix: oras-go client should fallback to docker config if no credentials specified

### LINKED ISSUE #20122: oras-go client should fallback to docker config if no credentials specified
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`: `2.11.0`

**Describe the bug**

I'm using https://github.com/argoproj/argo-cd/issues/17279 to authenticate to Google Artifact Registry as a helm registry using workload identity. Unlike https://github.com/argoproj/argo-cd/issues/10218, the other solution does not require installing ESO. This works in general since helm template can get creds via the $HOME/.docker/config.json. However, if you need to use `targetRevision: *` with your application, this ends up using the oras-go client here which is only configured for static credentials.

**To Reproduce**

1. Follow https://github.com/argoproj/argo-cd/issues/17279 to setup helm using an init container with a private repository
2. Add an application which uses helm with `targetRevision: *`
3. Permission denied

**Expected behavior**

oras-go uses the same creds as helm

### PR DESCRIPTION
**Pull request by tonyay163**
Fixes https://github.com/argoproj/argo-cd/issues/20122

I'm using https://github.com/argoproj/argo-cd/issues/17279 to authenticate to Google Artifact Registry as a helm registry using workload identity. Unlike https://github.com/argoproj/argo-cd/issues/10218, the other solution does not require installing ESO. This works in general since `helm template` can get creds via the `$HOME/.docker/config.json`. However, if you need to use `targetRevision: *` with your application, this ends up using the oras-go client here which is only configured for static credentials.

This change adds a backwards compatible change where if both the username and password is not configured, it will use the oras-go credentials package to get the docker config from the environment and use that to get credentials.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by tonyay163** [comment]
I tested this ETE and it seems to work well

**Comment by tonyay163** [comment]
@crenshaw-dev / @blakepettersson, sorry to ping directly but it looks like you've contributed this recently. I have a pretty simple fix here that I've been running for months now and would like to get it merged. It seems like a pretty small fix for an enhancement proposal.

**Review by gdsoumya — CHANGES_REQUESTED** [review]

File: util/helm/client_test.go
Comment (tonyay163) [reply util/helm/client_test.go]: hey, I think that's reasonable and have updated the test !

**Review by nitishfy — APPROVED** [review]
overall looks good to me!

**Comment by tonyay163** [comment]
thank you @nitishfy! I think we need @gdsoumya to approve or dismiss his requested changes

**Review by wanghong230 — APPROVED** [review]

**Review by gdsoumya — APPROVED** [review]

**Comment by tonyay163** [comment]
@gdsoumya, I'm waiting for this to release so I can upgrade my production envs, can you help me cherry-pick this to 2.11, 2.12 and 2.13 RCs?

**Comment by tonyay163** [comment]
thank you !

**Comment by eestolano** [comment]
Was this fix supposed to be in v2.13.0? It doesn't seem to be there.

**Pull request merged**
